### ISSUE TITLE
Add Quick Search Bar that seaches everywhere

### ISSUE THREAD

**Issue by rambda** [godotengine/godot-proposals#5586]
### Describe the project you are working on

It's a generically useful feature.

### Describe the problem or limitation you are having in your project

Finding something is not very convenient and very fast. It's convenient, but not ***very*** convenient, fast, but not ***very*** fast.
Currently it can be done by using default shortcuts and separate search bars. 
1. Focus file system filter: Ctrl+Alt+P. ~~BTW, I think Ctrl+P is more practical.~~
2. Find in files: Ctrl+Shift+F
3. Search Help: F1 / Alt+F1 ( now searches user scripts in Godot 4 )
4. Assign shortcuts to opening project/editor settings. Hard to memorize due to low frequency of use.
5. Command Palette: Ctrl+Shift+P
6. For evaluating math expressions, use shell or script languages the can be easily run in terminal.
7. For searching git, use terminal or source tree for searching git, which is the regular way.

**Edit: Godot does have a "Quick Open" (Shift+Alt+O) to search for and open any file in the file system. Although I think Ctrl + P as default is more common among most IDEs.**

Problems:
1. Too many shortcuts to memorize.
2. search boxes (LineEdit) are everywhere. There is no fixed position that a developer's eye expect to focus at.
3. ~~FileSystemDock is unpractical. When a file is found, the focus doesn't to the Tree; ui_down doesn't move the focus to the Tree either. You have to use Tab to move the focus to the tree, then navigate through it. I usually use the mouse... which is not convenient at all.~~
**Edit: Godot does have a "Quick Open" to do this.**
4. "Search Help" uses a filter to filter classes, constructors, methods, operators, signals, annotations, constants, properties, theme properties. but it's a drop-down menu, you have to click twice. There better be some way to unfold it
5. Indirect problem: Cannot open multiple FileSystemDock. I don't know why I usually want to have several folders open, but I just do that. If we can find files more quickly, maybe it minimize my panic.
6. Searching git in terminal requires some git experience.
7. it's better if everything can be done in the editor.

References:
1. Godot overall has the same design as most IDEs. using different shortcuts and separate search bars for searching different things. ~~But a "search filenames" bar is missing in Godot.~~
2. A "search filenames" bar is provided by many IDEs or text editors, like Visual Studio, VSCode, Sublime... I think at least this type of search bar is very practical.
3. A "search everywhere" bar is only provided by JetBrains IDEs. It's not a necessary feature, but a nice-to-have one. Developers need to remember just **_ONE_** shorcut. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a search/action bar that searches anything that exists as text in the project/editor, so that you can quickly jump to it.
You open this bar by a easy-to-press shorcut, like double tap Shift ( Jetbtains ).
You open this bar in a specific mode, like "search filenames only", by a shorcut, like Ctrl+P / Cmd+P ( VSCode / Sublime )
 
By "searching everything", it means searching for:
1. filename - to open the file in script editor / inspector / animation editor / shader editor / custom add-on tabs...
2. action - to run it. For example: export, orphan resource explorer, reload current project, reload saved scene, custom add-on actions...
3. autoload - to open the scene or script.
4. anything that "Search Help" provides: classes, constructors, methods, operators, signals, annotations, constants, properties, theme properties - to open the script/docs and jump to the corresponding line.
9. strings in node in packed scenes, or resources - to open the resource, or open the scene and jump to the node. For example, searches every `RichTextLabel.text` in saved scene, when you select the found item, the editor opens this scene and selects this node. (And you can choose not to search them in the filter.)
10. custom text files - to open the file and jump to this line. (And you can choose not to search them in the filter.)
11. project/editor setting - to jump to the entry.

**Good but not necessary:**
1. math expression evaluation - to get the result in drop-down. For example, input sqrt(9) + 1, get 4. Exactly what GDScript Expression class does.
2. git branches, commits, messages, tags - to open them in git plugin.

And you have filters like this:
![jetbrains](https://user-images.githubusercontent.com/23667522/195786776-98bc95e1-256b-4c64-af14-c05a4ab8853d.png)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

https://www.jetbrains.com/help/idea/searching-everywhere.html

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It will be used often. 
Thanks to https://gitlab.com/joao-pedro-braz/godot-finder , for 3.x only though.

### Is there a reason why this should be core and not an add-on in the asset library?

I think the main problem is it's hard to implement. But if it actually gets implemented, everyone will use it regularly. There's no reason for it to not be core.
Even if it should be a GDExtension, I don't think there is any better place to propose this.
And it's inefficient to be done in GDScript, because we don't have a GDScriptParser available in GDScript itself.

**Comment by RedMser** [comment]
Your proposed "search filenames" popup already exists with the "Quick Open" command (which also has modes "Quick Open Script" and "Quick Open Scene").

**Comment by RedMser** [comment]
> resources ??

@Zireael07 The basic "Quick Open" command searches for all files that are listed in the file explorer, including resources.
But I do agree that having better filtering options (e.g. "show script files extending `Theme`"), and also allow for searching file's contents, would be nice.

**Comment by me2beats** [comment]
I often search something in editor settings then in project settings. Maybe it would be cool to find in both at the same time

**Comment by rambda** [comment]
> Your proposed "search filenames" popup already exists with the "Quick Open" command (which also has modes "Quick Open Script" and "Quick Open Scene").

Wth? Has this feature always existed, even in 3.x? Thank you for telling me this.

**Comment by Calinou** [comment]
> Wth? Has this feature always existed, even in 3.x? Thank you for telling me this.

Yes, it's been around since 3.4 at least (<kbd>Shift + Alt + O</kbd>).

**Comment by RichardJECooke** [comment]
Shift-alt-o is very clumsy. Please make it F3, like Blender. Or even ? like Reaper.

It's so convenient to play any action from any menu in one place in these apps.

**Comment by Calinou** [comment]
> Shift-alt-o is very clumsy. Please make it F3, like Blender. Or even ? like Reaper.

<kbd>F3</kbd> opens the command palette in Blender; it is not a file searching tool. It's similar to Godot's command palette, which is opened using <kbd>Ctrl + Shift + P</kbd> (same shortcut as Visual Studio Code).
